web/app/view/CustomNumberField.js

```diff
@@ -20,30 +20,79 @@ Ext.define('Traccar.view.CustomNumberField', {
+                    if (!config.listeners) {
+                        config.listeners = {};
+                    }
+                    config.listeners.afterrender = function () {
+                        if (!this.units) {
+                            this.units = Ext.create({
+                                xtype: 'combobox',
+                                renderTo: 'numberUnitEl',
+                                store: 'TimeUnits',
+                                displayField: 'name',
+                                valueField: 'key',
+                                editable: false,
+                                numberField: this,
+                                value: 's',
+                                width: '70px',
+                                listeners: {
+                                    select: function () {
+                                        this.numberField.step = this.getStore().convertValue(1, this.getValue(), true);
+                                    }
+                                }
+                            });
+                        }
+                    };
+                    break;
+                default:
+                    break;
+            }
+            config.afterBodyEl = this.unitEl + unitName + '</div></div>';
             config.rawToValue = function (rawValue) {
-                return Ext.getStore('SpeedUnits').convertValue(rawValue, Traccar.app.getPreference('speedUnit', 'kn'), true);
+                if (this.units) {
+                    return this.units.getStore().convertValue(rawValue, this.units.getValue(), true);
+                } else {
+                    return this.parseValue(rawValue);
+                }
             };
             config.valueToRaw = function (value) {
-                return Ext.getStore('SpeedUnits').convertValue(value, Traccar.app.getPreference('speedUnit', 'kn'));
-            };
-        } else if (config.dataType === 'distance') {
-            config.beforeSubTpl = this.beforeEl;
-            unit = Traccar.app.getPreference('distanceUnit', 'km');
-            config.afterSubTpl = this.unitEl + Ext.getStore('DistanceUnits').findRecord('key', unit).get('name') + '</div></div>';
-            config.rawToValue = function (rawValue) {
-                return Ext.getStore('DistanceUnits').convertValue(rawValue, Traccar.app.getPreference('distanceUnit', 'km'), true);
-            };
-            config.valueToRaw = function (value) {
-                return Ext.getStore('DistanceUnits').convertValue(value, Traccar.app.getPreference('distanceUnit', 'km'));
+                if (this.units) {
+                    return this.units.getStore().convertValue(value, this.units.getValue(), false);
+                } else {
+                    return this.parseValue(value);
+                }
             };
+            if (config.units) {
+                config.step = config.units.getStore().convertValue(1, config.units.getValue(), true);
```

Comment:
Noticed that spiners (up/down) increase `value`, but not `rawValue`. That mean that step is in knots and meters, not current units.